### ISSUE TITLE
What is meant by Batch predictions for MNLI tasks

### ISSUE THREAD

**Issue by justachetan** [pytorch/fairseq#1167]
Hi! 

I am making use of the `roberta.large.mnli` model for obtaining entailment scores for sentences. 

I wanted to ask how exactly does the **batch predictions** example given in the README works. Specifically, I want to know that if I write the following code:

```
from fairseq.data.data_utils import collate_tokens
sentences = ['Hello world.', 'Another unrelated sentence.', 'Just another sentence.', 'DFTBA.']
batch = collate_tokens([roberta.encode(sent) for sent in sentences], pad_idx=1)
logprobs = roberta.predict('mnli', batch)
assert logprobs.size() == torch.Size([4, 3])
```

Then what does each row of the 4 x 3 tensor that is obtained represent? Even if someone could point me to the section of the codebase that could help me understand this, it would be a great help.

Thanks!

**Comment by justachetan** [comment]
I did eventually come across the `predict` function in `fairseq/fairseq/models/roberta/hub_interface.py` which I think is what is called in the code snippet above. But I still could not figure out how it handles the batch cases.

Another thing that seems weird is the need for padding the input with an extra zero in the beginning.

**Comment by justachetan** [comment]
@ngoyal2707 any updates on this?

**Comment by ngoyal2707** [comment]
Hey @justachetan, 

1) need of padding is to be able to create a batched input for the model when each sentences can have different number of tokens. So the input to the model is of shape [B x T x D]

2) [4x3] shape is [num_sentences x num_classes] since MNLI has 3 classes.

3) https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.glue.md has example of inference on `RTE`, it's not batched but it'd be easy to make it batched.

Let me know if you have more questions.

**Comment by justachetan** [comment]
But if the output if 4x3, between which two sentences is entailment indicated?

Also, entailment also depends on the order of the input sentences, how is that represented in the output?

It does not seem to be clear.

**Comment by justachetan** [comment]
I also tried taking the following sentences:

```sentences = ['Roberta is a heavily optimized version of BERT.', 'Roberta is not very optimized.', 'Roberta is based on BERT.']```

And I got the final prediction scores as:

```
tensor([[-2.6163, -0.3276, -1.5785],
        [-4.1085, -0.6111, -0.8192],
        [-2.0997, -0.4340, -1.4715]], device='cuda:0',
       grad_fn=<LogSoftmaxBackward>)
```

You will notice that the maximum elements in each row are the one with index 1. Which means that whatever pairs are being represented in each row are neutral to each other, but this is not the case when the examples are run separately (as indicated in your README)

**Comment by ngoyal2707** [comment]
So the input format of roberta 
```
<s> sent1 </s> </s> sent2 </s>
```
So for single sample prediction you need to use it like following:
```
tokens = roberta.encode(sent1, sent2)
prediction = roberta.predict('mnli', tokens).argmax().item()
```

**Comment by justachetan** [comment]
Okay, I think I am unable to get the point across, let me start again.

For a single pair prediction, the code looks something like this (taken from the README):
```
roberta = torch.hub.load('pytorch/fairseq', 'roberta.large.mnli')
roberta.eval()  # disable dropout for evaluation

# Encode a pair of sentences and make a prediction
tokens = roberta.encode('Roberta is a heavily optimized version of BERT.', 'Roberta is not very optimized.')
roberta.predict('mnli', tokens).argmax()  # 0: contradiction

```

In this case, the output of predict looks something like this:
```
tensor([[-6.4762, -4.0496, -0.0192]], device='cuda:0',
       grad_fn=<LogSoftmaxBackward>)
```

It would appear that the prediction here treats the first sentence as the premise and the second sentence as the hypothesis, and the results are given for this input in the above vector.

Now, if I want to do batched prediction, I would use this (taken from the README):
```
from fairseq.data.data_utils import collate_tokens
sentences = ['Roberta is a heavily optimized version of BERT.', 'Roberta is not very optimized.', 'Roberta is based on BERT.']
batch = collate_tokens([roberta.encode(sent) for sent in sentences], pad_idx=1)
logprobs = roberta.predict('new_task', batch)
assert logprobs.size() == torch.Size([3, 3])
```
The output in this case is:

```
tensor([[-2.6163, -0.3276, -1.5785],
        [-4.1085, -0.6111, -0.8192],
        [-2.0997, -0.4340, -1.4715]], device='cuda:0',
       grad_fn=<LogSoftmaxBackward>)
```

However, in this case, the output is a bit unclear. Because there are 3 sentences, the output can give entailment for six possible pairs of sentences, which are:

```
(sentences[0], sentences[1])
(sentences[0], sentences[2])
(sentences[1], sentences[2])
(sentences[1], sentences[0])
(sentences[2], sentences[0])
(sentences[2], sentences[1])
```

It is unclear, however, that which pairs have been predicted in the output. This is what I wanted to know. 

Also, the results seem to indicate that all the pairs that have been used for prediction are neutral to each other, which, going by your README, cannot be the case, as these same examples have been shown to have entailment as well as contradiction between them.

I hope this makes it more clear.

**Comment by ngoyal2707** [comment]
Gotcha, lemme try to write a batched version of pair classification task inference

**Comment by ngoyal2707** [comment]
```
import torch
from fairseq.data.data_utils import collate_tokens

roberta = torch.hub.load('pytorch/fairseq', 'roberta.large.mnli')
roberta.eval()

batch_of_pairs = [
    ['Roberta is a heavily optimized version of BERT.', 'Roberta is not very optimized.'],
    ['Roberta is a heavily optimized version of BERT.', 'Roberta is based on BERT.'],
    ['potatoes are awesome.', 'I like to run.'],
    ['Mars is very far from earth.', 'Mars is very close.'],
]

batch = collate_tokens(
    [roberta.encode(pair[0], pair[1]) for pair in batch_of_pairs], pad_idx=1
)

logprobs = roberta.predict('mnli', batch)
print(logprobs.argmax(dim=1))
```
Ouput
```
tensor([0, 2, 1, 0])
```
0 -> contradiction,
1 -> neutral
2 -> entailment

**Comment by justachetan** [comment]
Thank you so much! This clears things up 😄
